boi/mrp.py

```diff
@@ -0,0 +1,171 @@
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import fields, osv
+
+class mrp_production(osv.osv):
+
+    _inherit = 'mrp.production'
+
+    def _get_name(self, cr, uid, vals, context=None):
+        product_obj = self.pool.get('product.product')
+        order_obj = self.pool.get('sale.order')
+        if vals.get('active_id', False) and vals.get('active_ids', False):
+            for product in product_obj.browse(cr, uid, vals.get('active_ids')):
+                if product.ref_order_id:
+                    name = self.pool.get('ir.sequence').get(cr, uid, 'mrp.production')
+                    order = order_obj.browse(cr, uid, product.ref_order_id.id, context)
+                    if order.product_tag_id.name == 'BOI':
+                        name = 'BOI' + '-' + name
+                    else:
+                        name = 'NONBOI' + '-' + name
+                else:
+                    name = '/'
+        else:
+            name = '/'
+        return name
+
+    _defaults = {
+        'name': _get_name
+    }
+
+    def create(self, cr, uid, vals, context=None):
+        if context is None:
+            context = {}
+        if vals.get('name', '/') == '/':
+            vals['name'] = self.pool.get('ir.sequence').get(cr, uid, 'mrp.production')
+            order_obj = self.pool.get('sale.order')
+            if vals.get('order_id', False):
+                order = order_obj.browse(cr, uid, vals.get('order_id'), context=context)
+                if order.product_tag_id:
+                    if order.product_tag_id.name == 'BOI':
+                        vals['name'] = 'BOI' + '-' + vals.get('name')
+                    else:
+                        vals['name'] = 'NONBOI' + '-' + vals.get('name')
+        return super(mrp_production, self).create(cr, uid, vals, context=context)
+
+    def copy(self, cr, uid, id, default=None, context=None):
+        if default is None:
+            default = {}
+        if context is None:
+            context = {}
+        res = super(mrp_production, self).copy(cr, uid, id, default=default, context=context)
+        if res:
+            production = self.browse(cr, uid, res, context=context)
+            if production.order_id:
+                if production.order_id.product_tag_id:
```

Comment:
Always check for null
`if production.order_id and production.order_id.product_tag_id`